src/utils/api.ts

```diff
@@ -19,49 +22,118 @@ const handleCatch = (error: unknown) => {
-    return (await response.json()) as T;
+    const data = await response.json();
+    return data ? (data as T1) : undefined;
   } catch (error) {
     handleCatch(error);
   }
 };
 
-export const fetchBase = async <T1 extends { results: Array<T2> }, T2>(url: string) => {
-  try {
-    const response = await fetch(url);
-    const data = (await response.json()) as T1;
-    return data ? (data.results as Array<T2>) : undefined;
-  } catch (error) {
-    handleCatch(error);
-  }
+export const mapResults = async <T>(
+  results: Array<{
+    name: string;
+    url: string;
+  }>
+) => {
+  return await Promise.all(
+    results.map(({ url }) =>
+      fetch(url)
+        .then((res) => res.json())
+        .then((data) => data as T)
+    )
+  );
 };
 
-export const fetchDetails = async <T1 extends { url: string }, T2>(arr: Array<T1>) => {
-  try {
-    const responses = await Promise.all(arr.map((item) => fetch(item.url)));
-    const data = await Promise.all(responses.map((res) => res.json()));
-    return data as Array<T2>;
-  } catch (error) {
-    handleCatch(error);
+export const fetchAndMapPokemons = async <T extends PokemonBaseData>(url: string) => {
+  const data = (await genericFetch<T>(url)) || null;
+  const mappedResults = (await mapResults(data?.results || [])) as Array<PokemonDetails>;
+  return {
+    base: data,
+    mapped: mappedResults,
+  };
+};
+
+export const fetchAndMapMoves = async <T extends MoveBaseData>(url: string) => {
+  const data = (await genericFetch<T>(url)) || null;
+  const mappedResults = (await mapResults(data?.results || [])) as Array<PokemonMoveDetails>;
+  return {
+    base: data,
+    mapped: mappedResults,
+  };
+};
+
+export const fetchAndMapTypes = async <T extends TypeBaseData>(url: string) => {
+  const data = (await genericFetch<T>(url)) || null;
+  const mappedResults = (await mapResults(data?.results || [])) as Array<PokemonTypeDetails>;
+  return {
+    base: data,
+    mapped: mappedResults,
+  };
+};
+
+const fetchAndMapResults = async <
+  T extends {
+    currentPageResults?: Array<CardDetails>;
+    results: Array<{
+      name: string;
+      url: string;
+    }>;
   }
+>(
+  url: string
+) => {
+  const data = (await genericFetch<T>(url)) || null;
+  if (data) data.currentPageResults = await mapResults(data?.results || []);
+  return data;
 };
 
-export const fetchPokemonBase = async (url: string) => await fetchBase<PokemonData, Pokemon>(url);
-export const fetchMoveBase = async (url: string) => await fetchBase<MovesData, PokemonMove>(url);
-export const fetchTypeBase = async (url: string) => await fetchBase<TypesData, PokemonType>(url);
+export const fetchAndMapSortingData = {
+  pokemons: async (url: string) => {
+    const data = await genericFetch<BasePokemonsData>(url);
+    const mapped: PokemonDetails[] = await mapResults(data?.results || []);
+    if (data) {
+      const sortingData: BaseSearchPokemonsData = structuredClone(data);
+      sortingData.results = mapped;
```

Comment:
try/catch? check that error handling exists everywhere there on any request level. After adding it - check that it works by spoiling a link.

Reply:
done